### PR TITLE
sympy to sage conversion : Abstract function.

### PR DESCRIPTION
**Pull request by man74cio**
The solution proposed in the issue #12825 is implemented

### CODE REVIEW

**Comment by moorepants** [comment]
Can you add a unit test for this case? I think we have tests that run when sage is installed.

**Comment by man74cio** [comment]
Hello,
yes I can. could you please give me a reference page with instruction to add tests (with sage)?
I think to add the same test which describes the issue? it is ok for you?

**Comment by man74cio** [comment]
ok, I think I found were : sympy/external/tests/test_sage.py
Now I have to found how to write it. Any suggestion or reference ?

**Comment by moorepants** [comment]
I think that this:

```
from sympy import *
x,y = symbols('x y')
f = Function('f')
expr =  f(x,y)
expr._sage()
```

Is sufficient. It just shows that the error doesn't occur. You could add an assertion to check that the correct sage object was created too.

**Comment by man74cio** [comment]
Hello I added the unit test.
thanks 
Marco

File: sympy/external/tests/test_sage.py [hunk 1]
```diff
@@ -222,6 +222,17 @@ def test_undefined_function():
     assert bool(sf(sx) == f(x)._sage_())
     #assert bool(f == sympy.sympify(sf))
 
+def test_abstract_function():
+    x,y = sympy.symbols('x y')
+    f = sympy.Function('f')
+    expr =  f(x,y)
+    sexpr = expr._sage_()
+    invexpr = sexpr._sympy_()
+    assert type(sexpr) is sage.symbolic.expression.Expression, "converted expression %r is not sage expression" % sexpr
```
Comment (moorepants) [inline sympy/external/tests/test_sage.py]: I think `assert isinstance()` is probably a more standard way to do this kind of check.

**Comment by man74cio** [comment]
done

**Comment by moorepants** [comment]
Thanks. Now we wait to see if all tests pass.

**Comment by mforets** [comment]
Hi,
Is there something that can be done here to fix converting formal functions with no arguments? As in:

```python
sage: from sympy import Function
sage: f = Function('f')
sage: f(x)._sage_()  # fixed by this PR
f(x)
sage: f._sage_() 
...
TypeError: unbound method _sage_() must be called with f instance as first argument (got nothing instead)
```

But in Sage we have:

```python
sage: type(function('f'))
<class 'sage.symbolic.function_factory.NewSymbolicFunction'>
```

@rwst

**Comment by moorepants** [comment]
The test is failing on travis: https://travis-ci.org/sympy/sympy/jobs/248904009. Does the test pass locally for you?

**Comment by mforets** [comment]
the failing conversion `invexpr = sexpr._sympy_()` is addressed on [#20204](https://trac.sagemath.org/ticket/20204), so i guess it won't work here yet..

File: sympy/external/tests/test_sage.py [hunk 2]
```diff
@@ -222,6 +222,17 @@ def test_undefined_function():
     assert bool(sf(sx) == f(x)._sage_())
     #assert bool(f == sympy.sympify(sf))
 
+def test_abstract_function():
+    x,y = sympy.symbols('x y')
+    f = sympy.Function('f')
+    expr =  f(x,y)
+    sexpr = expr._sage_()
+    invexpr = sexpr._sympy_()
```
Comment (moorepants) [inline sympy/external/tests/test_sage.py]: Let's just leave this out for now. You can comment it out and leave a note to enable in the future.

**Comment by asmeurer** [comment]
FYI to whoever merges this, you need to manually check the Sage builds in Travis (they are in the "allowed failures", so they won't cause the build to red).

**Comment by man74cio** [comment]
I am sorry : I was testing with my personal branch of sage where I work on the inverse problem.
It is the ticket https://trac.sagemath.org/ticket/22802 . This work is not finished yet because I am trying to convert de derivative of abstract function.
I commented the test of invexpr = sexpr._sympy_()

**Comment by man74cio** [comment]
Hello, when this branch wille be merged in the trunk ?

**Comment by man74cio** [comment]
I need this information because it was asked me by the referee  related Sage ticket 
https://trac.sagemath.org/ticket/22802 .
**NOTE** as asked by the Sage referee the ticket sage#22802 was splitted into two separate tickets:
https://trac.sagemath.org/ticket/23496 containing the patch of sympy (with the modifications done in this ticket)
and 
https://trac.sagemath.org/ticket/22802 containing the modifications on sage to convert generic functions.

**Comment by asmeurer** [comment]
The sage tests are still failing on Travis.

**Comment by man74cio** [comment]
Could I give a look to these tests ?
which version of sage is used ?

**Comment by jksuom** [comment]
This is a log of sage tests: https://travis-ci.org/sympy/sympy/jobs/255622812

**Comment by man74cio** [comment]
I have tested with no problem:

 1) ipython in the sympy directory

 2)
 {{{
 import sys
 sys.path.append('/my_path_to_sage_8.0.beta2/src/sage')
 import mpmath
 run sympy/external/tests/test_sage.py
 }}}

I also downloaded the version 7.5.1 of sage / used python3.5 but I had no error

**Comment by man74cio** [comment]
Ok I could reproduce the error.
procedure:
1. Export env variables 
`
export  SAGE_ROOT=/tmp/SageMath/
export SAGE_LOCAL=$SAGE_ROOT/local
export PATH=$SAGE_ROOT/local/bin:$PATH
export LD_LIBRARY_PATH="$SAGE_ROOT/local/lib/openmpi:$SAGE_ROOT/local/lib/:$LD_LIBRARY_PATH"
`
2. in sympy directory : ipython
`
import sympy.external.tests.test_sage as test_sage
test_sage.test_abstract_function()
`
The error was in the use of the class "Expression" called directly as 
sage.symbolic.expression.Expression

Now I imported it as Expression, and no more errors (at least within my setting)

**Comment by man74cio** [comment]
Now the errors seem to be related to the build, but I think that my modifications cannot influence the build.

**Comment by asmeurer** [comment]
The sage tests in Travis passed.

**Pull request merged**